### PR TITLE
MutatingScope: implemented support for BitwiseNot operator

### PR DESCRIPTION
**Pull request by dktapps**

### CODE REVIEW

File: src/Analyser/MutatingScope.php [hunk 1]
```diff
@@ -569,6 +569,18 @@ private function resolveType(Expr $node): Type
...
+		if ($node instanceof \PhpParser\Node\Expr\BitwiseNot) {
+			if ($this->treatPhpDocTypesAsCertain) {
+				$exprType = $this->getType($node->expr);
+			} else {
+				$exprType = $this->getNativeType($node->expr);
+			}
+			if ($exprType instanceof StringType) {
+				return new StringType();
+			}
+			return $exprType->toInteger();
```
Comment (ondrejmirtes) [inline src/Analyser/MutatingScope.php:581]: Not covered cases:
* floats
* unions of constant strings

Best way to tackle this is to use TypeTraverse::map() which takes care of complex types easily...

File: src/Rules/Operators/InvalidUnaryOperationRule.php [hunk 2]
```diff
@@ -23,15 +23,24 @@ public function processNode(\PhpParser\Node $node, Scope $scope): array
...
 		}
 
 		if ($scope->getType($node) instanceof ErrorType) {
+
+			if ($node instanceof \PhpParser\Node\Expr\UnaryPlus) {
+				$operator = '+';
+			} elseif ($node instanceof \PhpParser\Node\Expr\UnaryMinus) {
+				$operator = '-';
+			} else {
+				$operator = '~';
```
Comment (ondrejmirtes) [inline src/Rules/Operators/InvalidUnaryOperationRule.php:38]: The changed rule should be in separate commit along with the related test changes.

File: src/Analyser/MutatingScope.php (hunk 1 above)
Comment (dktapps) [reply src/Analyser/MutatingScope.php:581]: Floats get coerced to integer for bitwise not. I hadn't considered the union types problem.

Comment (ondrejmirtes) [reply src/Analyser/MutatingScope.php:581]: Oh right, floats work (but a unit test would be nice).

File: src/Analyser/MutatingScope.php [hunk 3]
```diff
@@ -569,6 +569,18 @@ private function resolveType(Expr $node): Type
 			return new BooleanType();
 		}
 
+		if ($node instanceof \PhpParser\Node\Expr\BitwiseNot) {
+			if ($this->treatPhpDocTypesAsCertain) {
```
[the remark this answers is not in the document]
Comment (dktapps) [reply src/Analyser/MutatingScope.php:573]: I assumed it was since other similar operators need it as well. I'll get rid of it.

**Comment by ondrejmirtes** [comment]
Merged as https://github.com/phpstan/phpstan-src/commit/24246f3b134b13708b0aa259697b03e65bebb223, thank you.
